plugins/inventory/foreman.py

```diff
@@ -238,10 +296,205 @@ def _get_hostvars(self, host, vars_prefix='', omitted_vars=()):
+        params = {'input_values': self._fetch_params()}
+        try:
+            self.poll_interval = int(self.get_option('report').get('poll_interval'))
+        except Exception:
+            self.poll_interval = 10
+        ret = session.post(url, json=params)
+        if not ret:
+            raise Exception("Error scheduling inventory report on foreman. Please check foreman logs!")
+        url = "{0}/{1}".format(self.foreman_url, ret.json().get('data_url'))
+        response = session.get(url)
+        while response:
+            if response.status_code != 204:
+                break
+            sleep(self.poll_interval)
+            response = session.get(url)
+        if not response:
+            raise Exception("Error receiving inventory report from foreman. Please check foreman logs!")
+        else:
+            return response.json()
+
     def _populate(self):
+        if self._use_inventory_report():
+            self._populate_report_api()
+        else:
+            self._populate_host_api()
+
+    def _populate_report_api(self):
+        self.groups = dict()
+        self.hosts = dict()
+        try:
+            inventory_report_response = self._post_request()
+        except Exception as exc:
+            self.display.warning("Failed to use Reports API, falling back to Hosts API: {0}".format(exc))
+            self._populate_host_api()
+            return
+        self.group_prefix = self.get_option('group_prefix')
+        host_data = json.loads(inventory_report_response)
+        for host in host_data:
+            if not(host) or (host["name"] in self.cache.keys()):
+                continue
+            host_name = self.inventory.add_host(host['name'])
+            group_name = host.get('hostgroup_title', host.get('hostgroup_name'))
+            if group_name:
+                group_name = to_safe_group_name('%s%s' % (self.get_option('group_prefix'), group_name.lower().replace(" ", "")))
+                group_name = self.inventory.add_group(group_name)
+                self.inventory.add_child(group_name, host_name)
+
+            host_params = host.pop('host_parameters', {})
+            fact_list = host.pop('facts', {})
+
+            if self.get_option('legacy_hostvars'):
+                hostvars = self._get_hostvars(host)
+                self.inventory.set_variable(host_name, 'foreman', hostvars)
+            else:
+                omitted_vars = ('name', 'hostgroup_title', 'hostgroup_name')
+                hostvars = self._get_hostvars(host, self.get_option('vars_prefix'), omitted_vars)
+
+                for k, v in hostvars.items():
+                    try:
+                        self.inventory.set_variable(host_name, k, v)
+                    except ValueError as e:
+                        self.display.warning("Could not set host info hostvar for %s, skipping %s: %s" % (host, k, to_text(e)))
+
+            content_facet_attributes = host.get('content_attributes', {}) or {}
```

Comment:
```suggestion
            content_facet_attributes = host.get('content_attributes', {})
```

the `or {}` is not necessary, `get` will do this for us nicely

Reply:
I added this back. It wasn't handling the None case properly without this..

Reply:
So get returns None if the key was already set to None. But also as a default if we dont specify anything.
It should be `content_facet_attributes = host.get('content_attributes') or {}`.
Downside (, but the original version has the same problem): False, [], "" and similar is all translated to {}.

Reply:
Yeah. That's also what I thought after Samir said it didn't work as expected.

I am not sure if we write `host.get('content_attributes') or {}` now and someone will refactor that in a month to `host.get('content_attributes', {})` without thinking further. The current version is more explicit?